hcl2template/parser.go

```diff
@@ -177,7 +178,8 @@ func (p *Parser) decodeLocalVariables(f *hcl.File, cfg *PackerConfig) hcl.Diagno
 func (p *Parser) decodeConfig(f *hcl.File, cfg *PackerConfig) hcl.Diagnostics {
 	var diags hcl.Diagnostics
 
-	content, moreDiags := f.Body.Content(configSchema)
+	body := dynblock.Expand(f.Body, cfg.EvalContext())
+	content, moreDiags := body.Content(configSchema)
```

Comment:
Magic happens here and the actual code is in github.com/hashicorp/hcl/v2/ext/dynblock/ .

The rest is docs and tests.